binstar_client/utils/projects/__init__.py

```diff
@@ -0,0 +1,102 @@
+import inspect
+import logging
+import os
+from os.path import basename, isdir, join, relpath
+from binstar_client.errors import BinstarError
+from .filters import filters
+from .inspectors import inspectors
+
+log = logging.getLogger('binstar.projects.upload')
+
+
+class PFile(object):
+    def __init__(self, **kwargs):
+        self.fullpath = kwargs.get('fullpath', None)
+        self.basename = kwargs.get('basename', None)
+        self.relativepath = kwargs.get('relativepath', None)
+        self.size = kwargs.get('size', None)
+        self.populate()
+
+    def __str__(self):
+        if self.is_dir():
+            return self.relativepath
+        else:
+            return "[{}] {}".format(self.size, self.relativepath)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __eq__(self, other):
+        return self.fullpath == other.fullpath
+
+    def is_dir(self):
+        return os.path.isdir(self.fullpath)
+
+    def validate(self, validator):
+        if inspect.isfunction(validator):
+            return validator(basename=self.basename,
+                             relativepath=self.relativepath,
+                             fullpath=self.fullpath)
+        elif inspect.isclass(validator):
+            return validator(self)()
+        raise BinstarError("Invalid validator {}".format(validator))
+
+    def populate(self):
+        if self.size is None:
+            self.size = os.stat(self.fullpath).st_size
+        if self.basename is None:
+            self.basename = os.path.basename(self.fullpath)
+
+    def to_dict(self):
+        return {
+            'basename': self.basename,
+            'size': self.size,
+            'relativepath': self.relativepath
+        }
+
+
+def get_project_name(filename, args):
+    if os.path.isdir(filename):
+        return os.path.basename(filename)
+    else:
+        return os.path.splitext(os.path.basename(filename))[0]
```

Comment:
This logic will be in anaconda-project once I close https://github.com/Anaconda-Server/anaconda-project/issues/65 if you want to track that issue